Exception deserializing BasicDataset in CacheView

```stacktrace
Exception in thread "AWT-EventQueue-0" java.lang.NoClassDefFoundError: org/apache/log4j/Logger
	at com.inductiveautomation.ignition.common.XMLUtil.createSafeSAXParserFactory(XMLUtil.java:76)
	at com.inductiveautomation.ignition.common.XMLReaderPool.<init>(XMLReaderPool.java:44)
	at com.inductiveautomation.ignition.common.BasicDataset.<clinit>(BasicDataset.java:33)
	at java.base/java.io.ObjectStreamClass.hasStaticInitializer(Native Method)
	at java.base/java.io.ObjectStreamClass.computeDefaultSUID(ObjectStreamClass.java:1793)
	at java.base/java.io.ObjectStreamClass$1.run(ObjectStreamClass.java:290)
	at java.base/java.io.ObjectStreamClass$1.run(ObjectStreamClass.java:288)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:318)
	at java.base/java.io.ObjectStreamClass.getSerialVersionUID(ObjectStreamClass.java:287)
	at java.base/java.io.ObjectStreamClass.initNonProxy(ObjectStreamClass.java:592)
	at java.base/java.io.ObjectInputStream.readNonProxyDesc(ObjectInputStream.java:2051)
	at java.base/java.io.ObjectInputStream.readClassDesc(ObjectInputStream.java:1898)
	at java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2224)
	at java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1733)
	at java.base/java.io.ObjectInputStream.readArray(ObjectInputStream.java:2157)
	at java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1721)
	at java.base/java.io.ObjectInputStream$FieldValues.<init>(ObjectInputStream.java:2606)
	at java.base/java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2457)
	at java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2257)
	at java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1733)
	at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:509)
	at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:467)
	at io.github.paulgriffith.kindling.cache.CacheView.deserialize(CacheView.kt:270)
	at io.github.paulgriffith.kindling.cache.CacheView._init_$lambda$37(CacheView.kt:379)
	at java.desktop/javax.swing.DefaultListSelectionModel.fireValueChanged(DefaultListSelectionModel.java:224)
	at java.desktop/javax.swing.DefaultListSelectionModel.fireValueChanged(DefaultListSelectionModel.java:191)
	at java.desktop/javax.swing.DefaultListSelectionModel.setValueIsAdjusting(DefaultListSelectionModel.java:728)
	at java.desktop/javax.swing.plaf.basic.BasicTableUI$Handler.setValueIsAdjusting(BasicTableUI.java:992)
	at java.desktop/javax.swing.plaf.basic.BasicTableUI$Handler.mouseReleased(BasicTableUI.java:1205)
	at java.desktop/java.awt.AWTEventMulticaster.mouseReleased(AWTEventMulticaster.java:298)
	at java.desktop/java.awt.AWTEventMulticaster.mouseReleased(AWTEventMulticaster.java:297)
	at java.desktop/java.awt.Component.processMouseEvent(Component.java:6626)
	at java.desktop/javax.swing.JComponent.processMouseEvent(JComponent.java:3389)
	at java.desktop/java.awt.Component.processEvent(Component.java:6391)
	at java.desktop/java.awt.Container.processEvent(Container.java:2266)
	at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:5001)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2324)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4833)
	at java.desktop/java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4948)
	at java.desktop/java.awt.LightweightDispatcher.processMouseEvent(Container.java:4575)
	at java.desktop/java.awt.LightweightDispatcher.dispatchEvent(Container.java:4516)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2310)
	at java.desktop/java.awt.Window.dispatchEventImpl(Window.java:2780)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4833)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:773)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:722)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:716)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:97)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:746)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:744)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:743)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
Caused by: java.lang.ClassNotFoundException: org.apache.log4j.Logger
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
	... 61 more
```

--- a/src/main/kotlin/io/github/inductiveautomation/kindling/cache/CacheView.kt
+++ b/src/main/kotlin/io/github/inductiveautomation/kindling/cache/CacheView.kt
@@ -5,9 +5,11 @@ import com.formdev.flatlaf.extras.components.FlatPopupMenu
 import com.inductiveautomation.ignition.gateway.history.BasicHistoricalRecord
 import com.inductiveautomation.ignition.gateway.history.ScanclassHistorySet
 import com.jidesoft.swing.JideButton
+import io.github.inductiveautomation.kindling.cache.model.AbstractDataset
 import io.github.inductiveautomation.kindling.cache.model.AlarmJournalData
 import io.github.inductiveautomation.kindling.cache.model.AlarmJournalSFGroup
 import io.github.inductiveautomation.kindling.cache.model.AuditProfileData
+import io.github.inductiveautomation.kindling.cache.model.BasicDataset
 import io.github.inductiveautomation.kindling.cache.model.ScriptedSFData
 import io.github.inductiveautomation.kindling.core.Detail
 import io.github.inductiveautomation.kindling.core.DetailsPane
@@ -265,6 +267,8 @@ class CacheView(private val path: Path) : ToolPanel() {
         return AliasingObjectInputStream(inputStream()) {
             put("com.inductiveautomation.ignition.gateway.audit.AuditProfileData", AuditProfileData::class.java)
             put("com.inductiveautomation.ignition.gateway.script.ialabs.IALabsDatasourceFunctions\$QuerySFData", ScriptedSFData::class.java)
+            put("com.inductiveautomation.ignition.common.AbstractDataset", AbstractDataset::class.java)
+            put("com.inductiveautomation.ignition.common.BasicDataset", BasicDataset::class.java)
 //            put("com.inductiveautomation.ignition.gateway.alarming.journal.DatabaseAlarmJournal\$AlarmJournalSFData", AlarmJournalData::class.java)
 //            put("com.inductiveautomation.ignition.gateway.alarming.journal.DatabaseAlarmJournal\$AlarmJournalSFGroup", AlarmJournalSFGroup::class.java)
         }.readObject() as Serializable

--- a/src/main/kotlin/io/github/inductiveautomation/kindling/cache/model/Dataset.kt
+++ b/src/main/kotlin/io/github/inductiveautomation/kindling/cache/model/Dataset.kt
@@ -1,0 +1,79 @@
+package io.github.inductiveautomation.kindling.cache.model
+
+import com.inductiveautomation.ignition.common.Dataset
+import com.inductiveautomation.ignition.common.model.values.QualityCode
+import com.inductiveautomation.ignition.common.sqltags.model.types.DataQuality
+import java.io.ObjectInputStream
+
+@Suppress("PropertyName")
+abstract class AbstractDataset @JvmOverloads constructor(
+    @JvmField
+    protected var columnNames: List<String> = emptyList(),
+    @JvmField
+    protected var columnTypes: List<Class<*>> = emptyList(),
+    @JvmField
+    protected var qualityCodes: Array<Array<QualityCode>>? = null,
+) : Dataset {
+    @JvmField
+    protected var _columnNamesLowercase = columnNames.map { it.lowercase() }
+
+    override fun getColumnNames(): List<String> = columnNames
+    override fun getColumnTypes(): List<Class<*>> = columnTypes
+    override fun getColumnCount(): Int = columnNames.size
+    abstract override fun getRowCount(): Int
+    override fun getColumnIndex(columnName: String): Int = columnNames.indexOf(columnName)
+    override fun getColumnName(columnIndex: Int): String = columnNames[columnIndex]
+    override fun getColumnType(columnIndex: Int): Class<*> = columnTypes[columnIndex]
+
+    @Suppress("UNCHECKED_CAST")
+    private fun readObject(ois: ObjectInputStream) {
+        this._columnNamesLowercase = ois.readObject() as List<String>
+        columnNames = ois.readObject() as List<String>
+        columnTypes = ois.readObject() as List<Class<*>>
+        val qualities = ois.readObject()
+        qualityCodes = when {
+            qualities is Array<*> && qualities.isArrayOf<Array<QualityCode>>() -> qualities as Array<Array<QualityCode>>
+            qualities is Array<*> && qualities.isArrayOf<Array<DataQuality>>() -> (qualities as Array<Array<DataQuality>>).convertToQualityCodes()
+            else -> null
+        }
+    }
+
+    companion object {
+        @JvmStatic
+        private val serialVersionUID = -6392821391144181995L
+
+        private fun Array<Array<DataQuality>>.convertToQualityCodes(): Array<Array<QualityCode>> {
+            val columns = size
+            val rows = firstOrNull()?.size ?: 0
+            return Array(columns) { column ->
+                Array(rows) { row ->
+                    this[column][row].qualityCode
+                }
+            }
+        }
+    }
+}
+
+class BasicDataset @JvmOverloads constructor(
+    private val data: Array<Array<Any>> = emptyArray(),
+    columnNames: List<String> = emptyList(),
+    columnTypes: List<Class<*>> = emptyList(),
+    qualityCodes: Array<Array<QualityCode>>? = null,
+) : AbstractDataset(columnNames, columnTypes, qualityCodes) {
+    override fun getColumnNames(): List<String> = columnNames
+    override fun getColumnTypes(): List<Class<*>> = columnTypes
+    override fun getColumnCount(): Int = columnNames.size
+    override fun getRowCount(): Int = data.firstOrNull()?.size ?: 0
+    override fun getColumnIndex(columnName: String): Int = columnNames.indexOf(columnName)
+    override fun getColumnName(columnIndex: Int): String = columnNames[columnIndex]
+    override fun getColumnType(columnIndex: Int): Class<*> = columnTypes[columnIndex]
+    override fun getValueAt(rowIndex: Int, columnIndex: Int): Any = data[rowIndex][columnIndex]
+    override fun getValueAt(rowIndex: Int, columnName: String): Any = getValueAt(rowIndex, getColumnIndex(columnName))
+    override fun getQualityAt(rowIndex: Int, columnIndex: Int): QualityCode = QualityCode.Good
+    override fun getPrimitiveValueAt(rowIndex: Int, columnIndex: Int): Double = Double.NaN
+
+    companion object {
+        @JvmStatic
+        private val serialVersionUID = 3264911947104906591L
+    }
+}

